Add debug draw override to editor Game mode

Calinou opened the pull request:
This allows overriding the debug draw mode used in the root viewport.

For example, this can be used to temporarily view the running project in wireframe/overdraw mode, which can be used to better diagnose certain issues.

Note that certain debug draw modes have no effect in Mobile/Compatibility due to renderer limitations, but this limitation is already present in the editor. Since we reuse the 3D editor's code for the advanced menu, all the tooltips and disabled items according to the current renderer are still in place.

- This closes https://github.com/godotengine/godot-proposals/issues/12866.

## Preview

https://github.com/user-attachments/assets/14777a5a-46cc-4c4c-a8a3-7ee8f97d5579

## TODO

- [ ] Fix wireframe mode not working in Compatibility unless you call `RenderingServer.set_debug_generate_wireframes(true)` early enough in the project's lifecycle.
  - We could call this method automatically on startup when game embedding is enabled. Note that it has an impact on loading times and VRAM utilization, so maybe it should be an opt-in project setting.

Calinou committed:
Add debug draw override to editor Game mode

This allows overriding the debug draw mode used in the root viewport.

For example, this can be used to temporarily view the running project in
wireframe/overdraw mode, which can be used to better diagnose certain issues.

Note that certain debug draw modes have no effect in Mobile/Compatibility
due to renderer limitations, but this limitation is already present in the
editor.

syntaxerror247 commented:
Looks really good!

I think it'd be better to add a reset button next to the debug draw popup button and remove the 'Display Normal' option.

syntaxerror247 commented:
It'd be also really helpful to have a toggle button to show and hide collision shapes.
This PR doesn't need to add it though.

KoBeWi commented on editor/run/game_view_plugin.cpp:
```diff
@@ -734,6 +743,84 @@ void GameView::_camera_override_button_toggled(bool p_pressed) {
 	debugger->set_camera_override(p_pressed);
 }
 
+void GameView::_debug_draw_override_menu_id_pressed(int p_id) {
```
This handler can also be a static method in Node3DEditorViewport. The menus and callback can be arguments.

Calinou commented:
Rebased and tested again, it works as expected. 

> I think it'd be better to add a reset button next to the debug draw popup button and remove the 'Display Normal' option.

I'd prefer the UX of the dropdown to be 1:1 with the one found in the editor, so that we don't break users' muscle memory.

> It'd be also really helpful to have a toggle button to show and hide collision shapes. This PR doesn't need to add it though.

This requires https://github.com/godotengine/godot-proposals/issues/2072 to be implemented first.

> - This button should be invisible if 3D stuff is disabled in the editor feature profile (the same way as 3D button) - for 2D this button does nothing.

Done. I also fixed the icon being tinted while the dropdown is open.

> - Not sure, but could existing code be reused without duplication in game_view_plugin?

I refactored the code as per @KoBeWi's suggestions. There should be less duplicated code now.

> The debug draw options button shouldn't be between camera button and camera options. Maybe put it next to mute button?

I moved the debug draw mode dropdown to the right of the Mute button.

AThousandShips commented on editor/run/game_view_plugin.cpp:
```diff
@@ -1122,6 +1139,44 @@ void GameView::_notification(int p_what) {
+						break;
+					case Node3DEditorViewport::SupportedRenderingMethods::FORWARD_PLUS:
+						disabled = OS::get_singleton()->get_current_rendering_method() == "gl_compatibility" || OS::get_singleton()->get_current_rendering_method() == "mobile";
+						disabled_tooltip = TTR("This debug draw mode is not supported when using the Mobile or Compatibility renderers.");
```
```suggestion
						disabled_tooltip = TTR("This debug draw mode is not supported when using the Mobile or Compatibility renderer.");
```

YeldhamDev commented:
I wonder how much of this could be moved to `View3DController`, since now both the editor and runtime debug tools make use of it.
